### DISCUSSION TITLE
Docker image latest-prod runs "first run - install kimai" everytime the container starts

### DISCUSSION THREAD

**Discussion by wadedugas** [kimai/kimai#2427] [General]
Hi,

I created a docker compose file to use the Kimai docker image kimai/kimai2:latest-prod. How do I stop the kimai container from initializing Kimai every time the container starts?

Of course, this is a bit problematic for a production instance as it clears out all data on every restart ;-)

Here is the kimai service from my docker-copose.yml:

`  kimai:
    image: kimai/kimai2:latest-prod
    container_name: ${KIMAI_CONTAINER_NAME}
    environment:
      - APP_ENV=prod
      - TRUSTED_HOSTS=127.0.0.1,localhost,nginx,${HOSTNAME}
      - ADMINMAIL=${ADMINMAIL}
      - ADMINPASS=${ADMINPASS}
      - DB_TYPE=mysql
      - DATABASE_URL=mysql://${MYSQL_USER}:${MYSQL_PASSWORD}@mysqldb/kimaidb
      - MAILER_URL=${MAILER_URL}
      - MAILER_FROM=${MAILER_FROM}
    volumes:
      - public:/opt/kimai/public # << You must use a named volume, see note below
      - var:/opt/kimai/var
      # - ./ldap.conf:/etc/openldap/ldap.conf:z
      # - ./ROOT-CA.pem:/etc/ssl/certs/ROOT-CA.pem:z
    restart: unless-stopped
`

Thanks for your help!

**Comment by kevinpapst** [comment]
Re-post here: https://github.com/tobybatch/kimai2

**Reply by wadedugas** [reply]
Done!

https://github.com/tobybatch/kimai2/issues/220

**Reply by wadedugas** [reply]
This does seem to be an issue with the Docker image. I was able to create an image where it did not run the install at all using the Dockerfile from this repo minus the RUN commands found near the end. I did not get it working properly, but I was able to start the container and run the install commands manually and it did not install Kimai on every restart.

**Reply by kevinpapst** [reply]
The install command should definitely NOT remove any existing data. This is running in the container, to make sure that database updates are executed during a Kimai update.

**Reply by wadedugas** [reply]
I just figured out the issue and have a workaround.

The problem is a permission issue when startup.sh tries to write out the installed file:

function handleStartup() {
  \# first start?
  if ! [ -e /opt/kimai/installed ]; then 
    echo "first run - install kimai"
    /opt/kimai/bin/console -n kimai:install
    if [ ! -z "$ADMINPASS" ] && [ ! -a "$ADMINMAIL" ]; then
      /opt/kimai/bin/console kimai:create-user superadmin $ADMINMAIL ROLE_SUPER_ADMIN $ADMINPASS
    fi
  fi
  \# Add this here so it's always available, it would be lost between conatiner restarts.
  export KIMAI=$(/opt/kimai/bin/console kimai:version --short)
  **echo $KIMAI > /opt/kimai/installed**
  echo "Kimai2 ready"
}

The container user, www-data, does not have permission to create /opt/kimai/installed as /opt/kimai is only writable by root.

The workaround is to run the container once so it intializes Kimai. Then stop the container and create the file "installed" on your host system and add a volume mounting the file just created. Change the owner of the installed file to the container www-data user with `chown 82:82 installed` The modified compose file would look something like:


kimai:
    image: kimai/kimai2:latest-prod
    container_name: ${KIMAI_CONTAINER_NAME}
    environment:
      - APP_ENV=prod
      - TRUSTED_HOSTS=127.0.0.1,localhost,nginx,${HOSTNAME}
      - ADMINMAIL=${ADMINMAIL}
      - ADMINPASS=${ADMINPASS}
      - DB_TYPE=mysql
      - DATABASE_URL=mysql://${MYSQL_USER}:${MYSQL_PASSWORD}@mysqldb/kimaidb
      - MAILER_URL=${MAILER_URL}
      - MAILER_FROM=${MAILER_FROM}
    volumes:
      - public:/opt/kimai/public # << You must use a named volume, see note below
      - var:/opt/kimai/var
      **- ./installed:/opt/kimai/installed**
      # - ./ldap.conf:/etc/openldap/ldap.conf:z
      # - ./ROOT-CA.pem:/etc/ssl/certs/ROOT-CA.pem:z
    restart: unless-stopped

I see the latest version of the Dockerfile (which is not being used for the latest-prod image) has this statement `chown -R www-data:www-data /opt/kimai`. While this fixes this issue, is it really a good idea from a security standpoint or even necessary to make the www-data user the owner of the kimai root directory? If the container user owns the root directory of the application you might as well just run the container as root.

**Reply by kevinpapst** [reply]
Thanks for sharing, but again: please write that in the other repository. I have no Docker knowledge and I do NOT manage the images!

**Reply by wadedugas** [reply]
I already did: https://github.com/tobybatch/kimai2/issues/220#issuecomment-798902416

Just wanted to post it here as well in case it helps someone else.

**Reply by kevinpapst** [reply]
Awesome, thanks! I hope they see and fix it
